File: libnd4j/include/ops/declarable/generic/sparse/csr_edge_gather_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// csr_edge_gather_bp — backward pass of csr_edge_gather.
//
// Scatter-add upstream gradients back to node features:
//   dX[colIdx[e], f] += gradEdge[e, f]   for all e, f
//
// Multiple edges may reference the same source node → accumulate.
// dX is pre-zeroed before accumulation (OUTPUT_NULLIFIED; helpers also clear).
// On CUDA the helper uses sd_atomicAdd for the scatter; on CPU it is sequential.
//
// Inputs:
//   [0] colIdx    [nnz]    int   — source-node ids (same as csr_edge_gather forward)
//   [1] X         [n, F]   float — original node features (used ONLY for shape; values ignored)
//   [2] gradEdge  [nnz, F] float — upstream gradient w.r.t. edgeFeat
// IArgs: none
// Output:
//   [0] dX  [n, F]  float — gradient w.r.t. node features X (same shape as X)
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_csr_edge_gather_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_edge_gather.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(csr_edge_gather_bp, 3, 1, false, 0, 0) {
  auto colIdx   = INPUT_VARIABLE(0);  // [nnz]    int
  auto xArr     = INPUT_VARIABLE(1);  // [n, F]   float — shape source only, values unused
  auto gradEdge = INPUT_VARIABLE(2);  // [nnz, F] float

  REQUIRE_TRUE(colIdx->rankOf() == 1, 0,
               "csr_edge_gather_bp: colIdx must be rank-1, got %d", colIdx->rankOf());
  REQUIRE_TRUE(xArr->rankOf() == 2, 0,
               "csr_edge_gather_bp: X must be rank-2 [n, F], got %d", xArr->rankOf());
  REQUIRE_TRUE(gradEdge->rankOf() == 2, 0,
               "csr_edge_gather_bp: gradEdge must be rank-2 [nnz, F], got %d",
               gradEdge->rankOf());
  REQUIRE_TRUE(gradEdge->sizeAt(0) == colIdx->lengthOf(), 0,
               "csr_edge_gather_bp: gradEdge.shape[0] (%lld) must equal nnz (%lld)",
               (long long)gradEdge->sizeAt(0), (long long)colIdx->lengthOf());
  REQUIRE_TRUE(gradEdge->sizeAt(1) == xArr->sizeAt(1), 0,
               "csr_edge_gather_bp: gradEdge.shape[1] (%lld) must equal X.shape[1] (%lld)",
               (long long)gradEdge->sizeAt(1), (long long)xArr->sizeAt(1));

  auto dX = OUTPUT_NULLIFIED(0);  // [n, F] — same shape as X

  sd::ops::helpers::csr_edge_gather_bp(*colIdx, *xArr, *gradEdge, *dX);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(csr_edge_gather_bp) {
  // Input [1] is X[n, F]: use its shape for dX output shape.
  // Input [2] is gradEdge[nnz, F]: use its dtype for dX dtype.
  auto xShapeInfo        = inputShape->at(1);
  auto gradEdgeShapeInfo = inputShape->at(2);

  const sd::LongType n   = shape::sizeAt(xShapeInfo, 0);
  const sd::LongType F   = shape::sizeAt(xShapeInfo, 1);
  const sd::DataType dt  = ArrayOptions::dataType(gradEdgeShapeInfo);

  auto dxShape = ConstantShapeHelper::getInstance().createShapeInfo(dt, 'c', {n, F});
  return SHAPELIST(dxShape);
}

DECLARE_TYPES(csr_edge_gather_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})    // colIdx
      ->setAllowedInputTypes(1, {ALL_FLOATS})  // X (shape source)
      ->setAllowedInputTypes(2, {ALL_FLOATS})  // gradEdge
      ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_csr_edge_gather_bp)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CsrEdgeGatherBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Backward pass for {@link CsrEdgeGather}: scatter-add edge gradients to node features.
 *
 * <p>C++ op name: {@code csr_edge_gather_bp}
 * <ul>
 *   <li>Inputs:  colIdx[nnz] (INT), X[n, F] (node features — shape only, values unused),
 *                gradEdge[nnz, F] (upstream gradient)</li>
 *   <li>IArgs:   none — n and F are read from X's shape</li>
 *   <li>Output:  dX[n, F] — gradient w.r.t. node features X (same shape as X)</li>
 * </ul>
 *
 * <p>Gradient semantics:
 * <pre>
 *   dX[colIdx[e], f] += gradEdge[e, f]   for all e in [0, nnz), f in [0, F)
 * </pre>
 * Multiple edges may point to the same source node; their contributions are summed.
 * The output dX is pre-zeroed in the native kernel before accumulation.
 * X values are NOT read — X is passed purely so the C++ shape function can derive n and F
 * without relying on an IArg (which was unreliable in the SameDiff path).
 *
 * <p>This is a backward primitive: {@code doDiff} is not implemented.
 */
public class CsrEdgeGatherBp extends DynamicCustomOp {

    /** No-arg constructor required for op-registry reflection. */
    public CsrEdgeGatherBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param colIdx    1D [nnz] INT source-node ids
     * @param x         2D [n, F] node-feature matrix (shape source; values unused)
     * @param gradEdge  2D [nnz, F] upstream gradient w.r.t. edgeFeat
     */
    public CsrEdgeGatherBp(INDArray colIdx, INDArray x, INDArray gradEdge) {
        super(new INDArray[]{colIdx, x, gradEdge}, null);
        // No IArgs: n and F are read from x's shape in C++
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd        the SameDiff graph
     * @param colIdx    SDVariable [nnz] INT source-node ids
     * @param x         SDVariable [n, F] node features (shape source; values unused)
     * @param gradEdge  SDVariable [nnz, F] upstream gradient
     */
    public CsrEdgeGatherBp(SameDiff sd, SDVariable colIdx, SDVariable x, SDVariable gradEdge) {
        super(sd, new SDVariable[]{colIdx, x, gradEdge});
        // No IArgs: n and F are read from x's shape in C++
    }

    @Override
    public String opName() { return "csr_edge_gather_bp"; }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        // Inputs: [0]=colIdx(INT), [1]=X(FLOAT), [2]=gradEdge(FLOAT)
        // dX dtype = X dtype (input[1])
        return Collections.singletonList(dataTypes.get(1));
    }
}
